bird irradiance model implementation

tommarek opened the pull request.

tommarek committed:
bird irradiance model implementation

tommarek committed:
working total irradiance calculation, need to add tests

tommarek committed:
add precipitable water estimation, fix vectors, add test

tommarek committed:
fix fmt..

tommarek committed:
cloud sky irradiance model

tommarek committed:
add test

tommarek committed:
add tests

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+    time::minute,
+};
+
+const SOLAR_CONST: f64 = 1367.0; // W/m^2
```
Lepsi jmeno by bylo lepsi

tommarek replied:
https://en.wikipedia.org/wiki/Solar_constant

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+///
+/// # Returns
+/// * `Vector3<f64>` - three dimensional vector
+pub fn get_vector_from_azimuth_zenith(azimuth: &Angle, zenith_angle: &Angle) -> Vector3<f64> {
```
Vic by se mi libilo kdyby tato funkce byla definovana jako `get_vector_from_azimuth_elevation(azimuth, 90-zenith_angle)`

tommarek replied:
hotovo

bluecube replied:
Ted uz je to zas pouzite jen jednou, skoro je lepsi to pouziti nahradit tim 90-zenith_angle :)... co myslis?

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+///
+/// # Returns
+/// * `Area` - effective illuminated area
+pub fn get_effective_illuminated_area(
```
Neni tato funkce uz zbytecna?  `get_total_irradiance_on_tilted_surface` podle me dela vse na co by se tato funkce dala pouzit. Nebo se pletu?

bluecube replied:
Take ma velkou duplikaci s obsahem `get_projection`

tommarek replied:
odstranim

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+) -> f64 {
+    let sun_vector = get_vector_from_azimuth_zenith(solar_azimuth, solar_zenith);
+    let surface_vector = get_vector_from_azimuth_elevation(surface_azimuth, surface_angle);
+    sun_vector.dot(&surface_vector).max(0.0)
```
tady by se melo volat `get_illumination_coefficient`

tommarek replied:
pravda, opraveno :)

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+    Vector3::new(x, y, z).normalize()
+}
+
+// Get a dot product of the surface normal and the sun vector
```
Neni to uplne dot product, mas tam ten max. Chtelo by to lepsi popis

tommarek replied:
hotovo

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+// * `utc` - UTC time
+//
+// # Returns
+// * `f64` - ground albido value
```
albEdo

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+    anyhow::Ok(area)
+}
+
+// Returns typical ground albido value for a given month
```
albEdo

bluecube replied:
Taky by to mel byt docstring

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+//
+// # Returns
+// * `f64` - ground albido value
+pub fn get_typical_albedo(utc: &DateTime<Utc>) -> f64 {
```
Tohle je nezavisle na pozici? Ty cisla vypadaji prilis vysoke na zimu. Jestli je chapu dobre, tak ty 0.2 znamenaji ze je snih, a to u nas takhle dlouho rozhodne nebyva.

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+    let theta = air_temperature.get::<kelvin>() / 273.15;
+    let pw = 0.1
+        * (0.4976 + 1.5265 * theta + (13.6897 * theta - 14.9188 * (theta).powf(3.0)).exp())
+        * (216.7 * relative_humidity.get::<percent>() / (100.0 * air_temperature.get::<kelvin>())
```
Tady muzes pouzit jednotku `ratio` a zbavit se te stovky pod zlomkem.

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+//
+// # Returns
+// * `Length` - total precipitable water column
+pub fn get_total_precipitable_water(
```
Odkud je tenhle sileny vyraz? Muzes pridat link? (me prijde prakticke u takovychhle veci psat do komentare i cislo stranky v clanku nebo cislo rovnice, nebo tak)

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+//
+// # Returns
+// * `f64` - extraterrestrial solar radiation
+pub fn get_extraterrestrial_radiation(utc: &DateTime<Utc>) -> HeatFluxDensity {
```
🛸

bluecube replied:
Z clanku: "This value varies by ±3% as the earth orbits the sun". Ma vubec smysl tohle pocitat? Nebylo by lepsi tu proste placnout tu prumernou hodnotu?

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+
+// Calculate extraterrestrial solar radiation at any given time.
+// Average value for a day is 1367 W/m^2 which is corrected by a distance between the earth and the sun
+// Source: http://solardat.uoregon.edu/SolarRadiationBasics.html#Ref3
```
v odkazu je navic to `#Ref3`, myslim

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+// * `f64` - extraterrestrial solar radiation
+pub fn get_extraterrestrial_radiation(utc: &DateTime<Utc>) -> HeatFluxDensity {
+    let day_angle: Angle = Angle::new::<degree>(
+        (2.0 * std::f64::consts::PI / 365.0) * (f64::from(utc.ordinal()) - 1.0),
```
ten konstruktor mas ve stupnich, ale pouzivas tu `2 * PI` !

bluecube replied:
datetime ma `ordinal0`, nemusis odecitat jednicku

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+    //
+    // # Returns
+    // * `HeatFluxDensity` - sky diffuse irradiance on a tilted surface
+    fn get_sky_diffuse_irradiance_on_tilted_surface(
```
V tehle funkci porad bojujes proti tomu UOM, to nemusis. Je zbyetcne delat na kazdem miste `.get::<Foo>`, rusis tim tu rozmerovou kontrolu. Nebo to ma nejaky jiny duvod?

bluecube replied:
Neni to jen v tehle funkci, je to uplne vsude :)

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+                    .global_horizontal_irradiance
+                    .get::<watt_per_square_meter>())
+            .sqrt()
+                * (0.5 * surface_angle.get::<degree>()).sin().powf(3.0);
```
vytahujes ven z uom stupne, pak delas `f64::sin` v radianech.

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+
+#[derive(Debug, Clone, Copy)]
+pub struct Irradiance {
+    pub diffuse_horizontal_irradiance: HeatFluxDensity,
```
myslim ze by ty membery nemusely byt `pub`

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+    //
+    // # Returns
+    // * `ClearSkyIrradiance` - clear sky irradiance for given location and time
+    #[allow(clippy::too_many_arguments)]
```
clippy ma pravdu, mel bys mit asi nejaky struct s parametrama.

bluecube replied:
Odkud se vubec berou?

tommarek replied:
z meteorologickych dat :)

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        utc: &DateTime<Utc>,
+        lat: f64,
```
Tohle by mel byt angle

tommarek replied:
opraveno

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+    pub fn new(
+        utc: &DateTime<Utc>,
+        lat: f64,
+        lon: f64,
```
tohle taky

bluecube replied:
a nebo dohromady lat/lon v jedne strukture

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,813 @@
+            cloud_enhancement_estimate
+        };
+
+        let engerer2_coefficients = match averaging_period.get::<minute>() as u32 {
```
toto je velmi fuj. Co takhle ten algoritmus zafixovat a jednu periodu?

bluecube reviewed (changes requested):
Nasel jsem tam dve jasne chyby pouziti deg vs rad, taky tam mas skoro vsude takove hrozne komplikovane pouziti toho uom.

Moc teda celkove nechapu k cemu je to vsechno dobre, proc je tam tolik zpusobu jak ziskat irradiance a proc je to tak hrozne slozite :)

tommarek committed:
PR fixes

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,695 @@
+    Vector3::new(x, y, z).normalize()
+}
+
+// Get a coefficient value used for calculating effectively illuminated area
```
Tohle by mel byt docstring (`///`)

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,695 @@
+    get_illumination_coefficient(&sun_vector, &surface_vector)
+}
+
+// Get a coefficient value used for calculating effectively illuminated area
```
->docstring

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,695 @@
+// * `solar_zenith` - zenith angle of the sun (angle between the sun vector and the z axis)
+//
+// # Returns
+// * `f64` - dot product of the surface normal and the sun vector
```
tady porad mas ten "dot product", pritom je to spis ten "illumination coefficient"

tommarek committed:
fix docstrings + typos

bluecube committed:
Fixing errors

mostly these are deg/rad mismatches.
`apparent_solar_time`: in the linked python they add 12 hours instead of local solar time
engerer coefficients:  Coef 4 uses value in degrees in the python code

tommarek committed:
simplify irradiation models

tommarek committed:
cargo clippy fix

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,100 @@
+/// * `Ratio` - cloud cover factor
+fn could_factor(cloud_cover: Ratio) -> Ratio {
+    Ratio::new::<ratio>(
+        0.803 - 0.340 * cloud_cover.get::<ratio>() - 0.458 * cloud_cover.get::<ratio>().powi(2),
```
je tady potreba opoustet UOM?

`0.803 - 0.340 * cloud_cover- 0.458 * cloud_cover * cloud_cover,` nefunguje?
Pokud ne, tak co `Ratio::new::<ratio>(0.803) - 0.340 * cloud_cover- 0.458 * cloud_cover * cloud_cover,` ?

bluecube commented on MPCHomeControl/src/tools/sun.rs:
```diff
@@ -0,0 +1,100 @@
+    longitude: Angle,
+    datetime: &DateTime<Utc>,
+    cloud_cover: Ratio,
+    surface_angle: Angle,
```
napadlo me prejmenovat na `surface_angle_from_horizontal`  (nebo from vertical?)

bluecube reviewed (commented):
Mam dve super minor poznamky, pokud se ti nechce tak neni ani jedna z nich potreba. Predpokladam ze to ze nemas indirect ambient light je naschval. Vis jak moc silny ten efekt je, ze se ho nevyplati modelovat?

tommarek committed:
PR changes
